components/external_storage/export/src/dylib.rs

```diff
@@ -0,0 +1,308 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::request::{
+    anyhow_to_io_log_error, file_name_for_write, restore_receiver, restore_sender, write_receiver,
+    write_sender, DropPath,
+};
+use anyhow::Context;
+use external_storage::ExternalStorage;
+use futures_io::AsyncRead;
+use kvproto::backup as proto;
+use lazy_static::lazy_static;
+use once_cell::sync::OnceCell;
+use protobuf::{self, Message};
+use slog_global::{error, info, warn};
+use std::io::{self, ErrorKind};
+use std::sync::{Arc, Mutex};
+use tikv_util::time::Limiter;
+use tokio::runtime::{Builder, Runtime};
+
+#[cfg(feature = "prost-codec")]
+pub use kvproto::backup::storage_backend::Backend;
+#[cfg(feature = "protobuf-codec")]
+pub use kvproto::backup::StorageBackend_oneof_backend as Backend;
+
+struct ExternalStorageClient {
+    backend: Backend,
+    runtime: Arc<Runtime>,
+    blob_storage: Arc<Box<dyn ExternalStorage>>,
```

Comment:
why should we keep a blob_storage in side this client? It seems the only use of the blob storage is to build the file path of shared data

Reply:
yes, it is only used for metadata. I agree it should be removed. This code is tested to work but I still consider it experimental. We can spend more time cleaning things up if we commit to using this approach. Properly organizing the build for the cdylib is the first things I would look at cleaning up.

Reply:
I went ahead and removed the blob_storage reference since it was an easy change to make.